src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -573,12 +573,76 @@ private Task InvokeAllExceptionFiltersAsync()
 
                         _logger.ActionMethodExecuting(_actionExecutingContext, arguments);
 
-                        var actionReturnValue = await ControllerActionExecutor.ExecuteAsync(
-                            _executor,
-                            _controller,
-                            arguments);
+                        var returnType = _executor.MethodReturnType;
 
-                        result = CreateActionResult( actionMethodInfo.ReturnType, actionReturnValue);
+                        if (returnType == typeof(void))
+                        {
+                            _executor.Execute(_controller, arguments);
+                            result = new EmptyResult();
+                        }
+                        else if (returnType == typeof(Task))
+                        {
+                            await (Task)_executor.Execute(_controller, arguments);
+                            result = new EmptyResult();
+                        }
+                        else if (returnType == typeof(IActionResult))
+                        {
+                            result = (IActionResult)_executor.Execute(_controller, arguments);
+                            if (result == null)
+                            {
+                                throw new InvalidOperationException(
+                                    Resources.FormatActionResult_ActionReturnValueCannotBeNull(typeof(IActionResult)));
+                            }
+                        }
+                        else if (_executor.TaskGenericType == typeof(IActionResult))
+                        {
+                            result = await (Task<IActionResult>)_executor.Execute(_controller, arguments);
+                            if (result == null)
+                            {
+                                throw new InvalidOperationException(
+                                    Resources.FormatActionResult_ActionReturnValueCannotBeNull(typeof(IActionResult)));
+                            }
+                        }
+                        else if (_executor.IsTypeAssignableFromIActionResult)
```

Comment:
> Yes, there is one difference. if the _executor.TaskGenericType == typeof(IActionResult) , we still execute it in sync because we can cast to Task and await on it. We can get rid of the returntype == typeof(IActionResult) as this is covered by IsTaskAssignableFromIActionResult and is sync

Yeah I'd like to see the case for sync `== typeof(IActionResult)` collapsed with the case for sync `typeof(IActionResult).IsAssignableFrom(...)`. Other than that :shipit: